### PR TITLE
Added automated release process to project pipeline.

### PR DESCRIPTION
**Pull request by neil-k-zero**
- Project pipeline moved to azure-linux executor allowing public IP
access for workflow tests.

- Reconfigured bin/publish to perform releases and promotions
of contianer images based on optional flag parameters.

- Added additional functions in bin/build_utils for the automated releases.

- Maintained legacy bin/build_utils functions that are currently no longer in
use, but might provide useful in the future.

- Added submodule for conjur-opentelemetry-tracer upstream dependency.

- Moved temp directory creation to /bin/test-workflow/start. The temp directory
must be created by the jenkins user to prevent permission issues within
the directory. This also removes the possibility of a race condition when parallel
test stages are running and allows for developers to skip tests
during the development process.

### Desired Outcome

Add project to automated release process.

### Implemented Changes

- Jenkins executor is now an Azure instance to allow for public facing IP access.
 
- Version file is automatically generated from the CHANGELOG, this is utilized for automated
releases.

- Github pre-release are automatically generated with the option to promote said pre-release
to a GitHub release for customers.

- Docker Software Bill of Materials is generated on releases and is stored as an artifact. This can
be utilized to regenerate the go.mod file.

- Container images are tagged as edge for pre-releases.

- Images that are promoted (released) are tagged as latest. 

### Connected Issue/Story

N/A

#### Changelog

- [x] The CHANGELOG has been updated, or
- [ ] This PR does not include user-facing changes and doesn't require a
  CHANGELOG update

#### Test coverage

- [ ] This PR includes new unit and integration tests to go with the code
  changes, or
- [x] The changes in this PR do not require tests

#### Documentation

- [ ] Docs (e.g. `README`s) were updated in this PR
- [ ] A follow-up issue to update official docs has been filed here: [insert issue ID]()
- [x] This PR does not require updating any documentation

#### Behavior

- [ ] This PR changes product behavior and has been reviewed by a PO, or
- [ ] These changes are part of a larger initiative that will be reviewed later, or
- [x] No behavior was changed with this PR

#### Security

- [ ] Security architect has reviewed the changes in this PR,
- [ ] These changes are part of a larger initiative with a separate security review, or
- [x] There are no security aspects to these changes

### CODE REVIEW

File: Jenkinsfile [hunk 1]
```diff
@@ -129,7 +179,7 @@ pipeline {
             stage('Test app in GKE') {
               steps {
                 sh '''
-                  HOST_IP="$(curl http://169.254.169.254/latest/meta-data/public-ipv4)";
+                  HOST_IP="$(curl https://checkip.amazonaws.com)";
```
Comment (diverdane) [inline Jenkinsfile:182]: Ooh, this is a cool service!

File: bin/publish [hunk 2]
```diff
@@ -1,54 +1,103 @@
...
+  echo "Promote Usage: $0 --source <VERSION> --target <VERSION>"
+  echo " --source <VERSION>: specify version number of local image"
+  echo " --target <VERSION>: specify version number of remote image"
+}
 
-readonly REGISTRY="cyberark"
+while [[ $# -gt 0 ]]; do
+  case "$1" in
+  --source)
+    SOURCE_TAG=$2
```
Comment (diverdane) [inline bin/publish:19]: Nit: Should be quotes around `$2` here and on line 23 below. (I know we haven't been terribly consistent on this. :/ )
(Ref: https://google.github.io/styleguide/shellguide.html#s5.7-quoting)

**Review by diverdane — DISMISSED** [review]
Looks good to me!!! The auto-release code looks good to my eyes, but I'm not an expert.

With the addition of the conjur-opentelemetry-tracer as a submodule, should we consider adding some steps to CONTRIBUTING.md to handle submodule init and update and such? I think there are a few commands that devs will need to be aware of, if I remember submodule management correctly?  We can probably add this to our C&I backlog, if it needs to be done.

File: bin/publish (hunk 2 above)
Comment (neil-k-zero) [reply bin/publish]: Fixed, thanks.

**Comment by andytinkham** [comment]
I heard back from Snyk support - Snyk does support the replace command in go.mod, but doesn't yet support relative paths for it. They put in a feature request, but who knows when that will be available. 

Can we rework the go.mod replace command to a different path somehow?

**Comment by neil-k-zero** [comment]
@andytinkham Jenkins has a dynamic path to the repo based on the executor and the branch that's being built at the time which is why a relative path is being used in the go.mod. Not sure if there's a way around this.

**Comment by neil-k-zero** [comment]
> @neil-k-zero -I don't know go inclusion well enough to second guess that, so I guess the answer is we need to back off the Snyk scans and PR integrations in this repo, and find another way to scan the go modules while we wait for Snyk to add support. I don't have any visibility on the feature request support said they'd submit for me, so no idea on timing.
> 
> I have no idea if it's possible, but could we maybe use pwd somehow? Don't know if Snyk would like that any better if the '.' doesn't resolve though.

The go.mod file syntax is restricted to the absolute path or relative path.

File: CHANGELOG.md [hunk 3]
```diff
@@ -6,6 +6,10 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 
 ## [Unreleased]
 
+## [0.23.2] - 2022-03-23
+### Changed
+- Update to automated release process
```
Comment (szh) [inline CHANGELOG.md:11]: Can you add a link to the PR here?

Comment (neil-k-zero) [reply CHANGELOG.md]: Done!

File: go.mod [hunk 4]
```diff
@@ -1,10 +1,9 @@
 module github.com/cyberark/conjur-authn-k8s-client
 
-go 1.17
+go 1.16
```
Comment (neil-k-zero) [inline go.mod:3]: Downgraded to correlate with the Dockerfile version. Needs further discussion.

File: Dockerfile.helm-unit-test [hunk 5]
```diff
@@ -21,6 +21,9 @@ RUN wget https://github.com/mikefarah/yq/releases/latest/download/yq_linux_amd64
 RUN mkdir -p /conjur-authn-k8s-client
 WORKDIR /conjur-authn-k8s-client
 
+# Workaround for CVE-2022-24765 when running git inside a docker container
```
Comment (neil-k-zero) [inline Dockerfile.helm-unit-test:24]: Not sure if this is the only way around this issue as of yet but due to a git security fix, the container needs to explicitly run a safe.directory command before running.

File: go.mod (hunk 4 above)
Comment (rpothier) [reply go.mod:3]: ok, we can follow up after this PR gets in.

**Review by rpothier — APPROVED** [review]
LGTM! Thanks for getting this in.

**Pull request merged**